webapp/channels/src/packages/mattermost-redux/src/reducers/entities/files.ts

```diff
@@ -188,9 +188,96 @@ function filePublicLink(state: {link: string} = {link: ''}, action: GenericActio
     }
 }
 
+function filePreviews(state: Record<string, FilePreviewInfo[]> = {}, action: GenericAction) {
+    switch (action.type) {
+    case FileTypes.RECEIVED_FILE_PREVIEWS:
+        return {
+            ...state,
+            [action.pendingPostId]: action.data,
+        };
+
+    case FileTypes.UPDATE_FILE_UPLOAD_PROGRESS: {
+        const nextState = {...state};
+        const entries = Object.entries(nextState);
+        for (let entryIndex = 0; entryIndex < entries.length; entryIndex++) {
+            const entry = entries[entryIndex];
+            const filePreviewInfoIndex = entry[1].findIndex((f) => f.clientId === action.data.clientId);
+            if (filePreviewInfoIndex === -1) {
+                continue;
+            }
+            const pendingPostId = entry[0];
+            const filePreviewInfo = nextState[pendingPostId][filePreviewInfoIndex];
+            nextState[pendingPostId] = nextState[pendingPostId].map((f) => {
+                if (f.clientId !== filePreviewInfo.clientId) {
+                    return f;
+                }
+                return {
+                    ...f,
+                    percent: action.data.percent,
+                };
+            });
+            break;
+        }
+        return nextState;
+    }
+
+    case FileTypes.REMOVE_FILE_PREVIEWS: {
+        const nextState = {...state};
+        Reflect.deleteProperty(nextState, action.pendingPostId);
+        return nextState;
+    }
+
+    case FileTypes.UPLOAD_FILES_CANCEL: {
+        const entry = Object.entries(state).find((entry) => entry[1].some((f) => f.clientId === action.clientId));
+        if (!entry) {
+            return state;
+        }
+        return {
+            ...state,
+            [entry[0]]: entry[1].filter((f) => f.clientId !== action.clientId),
```

Comment:
A bit more nitpicking here, but would you mind redeclaring `entry[0]` and `entry[1]` as something like `id` and `uploadsInProgress` to make it more obvious what's going on here?